// @neo4j/graphql/dist/classes/Node.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const pluralize_1 = __importDefault(require("pluralize"));
const GraphElement_1 = require("./GraphElement");
const lower_first_1 = require("../utils/lower-first");
class Node extends GraphElement_1.GraphElement {
    constructor(input) {
        super(input);
        this.relationFields = input.relationFields;
        this.connectionFields = input.connectionFields;
        this.cypherFields = input.cypherFields;
        this.otherDirectives = input.otherDirectives;
        this.unionFields = input.unionFields;
        this.interfaceFields = input.interfaceFields;
        this.interfaces = input.interfaces;
        this.objectFields = input.objectFields;
        this.exclude = input.exclude;
        this.nodeDirective = input.nodeDirective;
        this.fulltextDirective = input.fulltextDirective;
        this.auth = input.auth;
        this.queryOptions = input.queryOptionsDirective;
    }
    // Fields you can set in a create or update mutation
    get mutableFields() {
        return [
            ...this.temporalFields,
            ...this.enumFields,
            ...this.objectFields,
            ...this.scalarFields,
            ...this.primitiveFields,
            ...this.interfaceFields,
            ...this.objectFields,
            ...this.unionFields,
            ...this.pointFields,
        ];
    }
    /** Fields you can apply auth allow and bind to */
    get authableFields() {
        return [
            ...this.primitiveFields,
            ...this.scalarFields,
            ...this.enumFields,
            ...this.unionFields,
            ...this.objectFields,
            ...this.temporalFields,
            ...this.pointFields,
            ...this.cypherFields,
        ];
    }
    get constrainableFields() {
        return [...this.primitiveFields, ...this.temporalFields, ...this.pointFields];
    }
    get uniqueFields() {
        return this.constrainableFields.filter((field) => field.unique);
    }
    get plural() {
        var _a;
        const pluralValue = ((_a = this.nodeDirective) === null || _a === void 0 ? void 0 : _a.plural) ? this.nodeDirective.plural : (0, pluralize_1.default)(this.name);
        return (0, lower_first_1.lowerFirst)(pluralValue);
    }
    getLabelString(context) {
        var _a;
        return ((_a = this.nodeDirective) === null || _a === void 0 ? void 0 : _a.getLabelsString(this.name, context)) || `:${this.name}`;
    }
    getLabels(context) {
        var _a;
        return ((_a = this.nodeDirective) === null || _a === void 0 ? void 0 : _a.getLabels(this.name, context)) || [this.name];
    }
    getMainLabel() {
        var _a;
        return ((_a = this.nodeDirective) === null || _a === void 0 ? void 0 : _a.label) || this.name;
    }
}
exports.default = Node;
